guides/shipments/custom-shipping-calculators.md

```diff
@@ -0,0 +1,68 @@
+# Custom shipping calculators 
+
+This article provides context about creating custom shipping calculators in the
+case that the provided calculators do not meet your store's needs.
+
+## Pre-configured calculators
+
+Before developing a custom calculator, you should make sure that the calculator
+you need doesn't already exist in Solidus or one of its extensions.
+
+Solidus comes with a set of default calculators that account for typical
+shipping scenarios:
+
+- [Flat percent](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/calculator/shipping/flat_percent_item_total.rb)
+- [Flat rate (per order)](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/calculator/shipping/flat_rate.rb)
+- [Flat rate per package item](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/calculator/shipping/per_item.rb)
+- [Flexible rate per package item](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/calculator/shipping/flexi_rate.rb)
+- [Price sack](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/calculator/shipping/price_sack.rb)
+
+If the calculators that come with Solidus are not enough for your needs, you
+might want to use an extension like
+[`solidus_active_shipping`][solidus-active-shipping] that provides additional 
+API-based rate calculation functionality for common carriers like UPS, USPS, and
+FedEx. Alternatively, you could develop your own custom calculator. 
+
+[solidus-active-shipping]: solidus-active-shipping-extension.md
+
+## Custom calculator requirements
+
+Your calculator should accept an array of `LineItem` objects that are associated
```

Comment:
The calculator takes a `Spree::Stock::Package` object, not an Array of line items. That object has access to the `package.order` and through that to all other stuff in the order. 

However, when dealing with the contents of the package (that really is a shipment), as a developer you *must* deal with the `package.contents` array. If your store does split shipments, you will otherwise quote the entire order when you only want to quote the contents of one shipment of many.